### ISSUE TITLE
[Feature request, discussion] Uppercase schema in QR-encoded BIP21strings

### ISSUE THREAD

**Issue by harding** [bitcoin-core/gui#145]
Bitcoin Core 0.19 through 0.21 (at least) uppercases the bech32 address in BIP21 strings so that they will produce less complex QR codes that can sometimes take up less space or be quicker to scan accurately.

![2020-12-08-09_32_37_500950206](https://user-images.githubusercontent.com/61096/101497401-efdb2380-3938-11eb-90d0-cbed6e552eb1.png)

However, RFC3986 allows treating URI schemas as case insensitive, so it should be possible to also uppercase the leading `BITCOIN` (and the separator `:` is part of the QR code uppercase set).  RFC3986 [says](https://tools.ietf.org/html/rfc3986#section-3.1):

> An implementation should accept uppercase letters as equivalent to lowercase in scheme names (e.g., allow "HTTP" as well as "http") for the sake of robustness but should only produce lowercase scheme names for consistency.

[BIP21](https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki#abnf-grammar) agrees:

> The scheme component ("bitcoin:") is case-insensitive, and implementations must accept any combination of uppercase and lowercase letters. 

One **important** downside of this approach is that not all wallets currently implement RFC3986 according to specification.  See https://github.com/btcpayserver/btcpayserver/issues/2099 for at least one instance of this.  It's my hope that the recent attention the BTCPay team was able to bring to this issue will help get any other wallets with this issue to upgrade in the near future, allowing Bitcoin Core and others to use more efficient uppercase schema strings with bech32 addresses.

**Comment by pavlenex** [comment]
We're investigating the adoption and wallet support of the upper-cased Bech32 and community is reaching out to wallets that aren't supporting the uppercase. You can follow on the progress we've made in[ this issue.](https://github.com/btcpayserver/btcpayserver/issues/2110)

**Comment by harding** [comment]
@pavlenex that's awesome, thank you all!

**Comment by Sjors** [comment]
For address with a label the `bitcoin:` to `BITCOIN:` would be quite marginal, and perhaps not worth the risk of incompatibility with some ancient wallet. On the other hand: why on earth would someone send coins _from_ Bitcoin Core _to_ an ancient wallet?

**Comment by harding** [comment]
@Sjors I agree the benefit is fairly marginal (without checking specs, I think uppercase vs. alphanumeric saves like 3 bits per character plus up to 22 bits for mode change, so best case savings is like `46 bits = 3*8+22`.  For BIP21 strings with a P2WPKH address and an amount of 1.00000000 BTC (no label or extension strings), that's best case savings of about 13.5%).

That said, I think interactions with "ancient" wallets are already going to require the user produce a non-default address type:

- For many wallets last updated more than a year or two ago, bech32 isn't supported, so the Bitcoin Core user is going to need to generate a legacy address anyway.  We can keep the `bitcoin:` schema lowercase in that case since we already keep the base58check string case sensitive.

- For any wallet released as of this writing, it won't support paying to the segwit v1+/taproot addresses that are likely to be deployed, so the Bitcoin Core user is going to need to generate an earlier address type in that case as well.

If users with old wallets are going to need to deal with address incompatibilities for other reasons, I think it's ok to risk asking them to deal with backwards incompatibility due to URI parsing problems too.

**Comment by Rspigler** [comment]
@ChristopherA BCR-2020-005 supports uppercase and lowercase, so should be compliant with RFC3986?

**Comment by ChristopherA** [comment]
[bcr-2020-05](https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-005-ur.md) is specifically designed to work with multiple transports, and is not case-significant.

We suggest lowercase for readability, but for use in QRs to take advantage of the QR-native compression, you need all uppercase. Both forms are completely URI compatible. See [bcr-2020-003](https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-003-uri-binary-compatibility.md) for more details on meeting the requirements of RFC3986 and other requirements.

/cc @wolfmcnally
